plugin/metadata/metadataer.go

```diff
@@ -0,0 +1,58 @@
+package metadata
+
+import (
+	"context"
+
+	"github.com/miekg/dns"
+)
+
+// Metadataer interface needs to be implemented by each plugin willing to provide
+// metadata information for other plugins.
+// Note: this method should work quickly, because it is called for every request
+// from the metadata plugin.
+type Metadataer interface {
+	// Metadata is expected to return map with metadata information which can be
+	// later retrieved from context by any other plugin. It may return empty
+	// map if no metadata needs to be published.
+	Metadata(context.Context, dns.ResponseWriter, *dns.Msg) map[string]interface{}
+}
+
+// MD is metadata information storage
+type MD map[string]interface{}
+
+// MDKeyType defines the type of key that is used to save metadata into the context
+type mDKeyType string
+
+// MDKey specifies
+const mDKey mDKeyType = "md-key"
+
+type metadataKey struct{}
```

Comment:
you can still use this below instead of the string - it's fine now to be unexported because you have the FromContext function.

Reply:
It was my first implementation and I forgot to remove it when moving to "const mDKey mDKeyType". But I can switch back to metadataKey{} if you think it looks better.

Reply:
Yes, please do.